File: swift-bridge/Sources/CoreTextBridge/Line.swift
import CoreText
import Foundation

private func truncationType(_ raw: UInt32) -> CTLineTruncationType {
    switch raw {
    case 1: return kCTLineTruncationEnd
    case 2: return kCTLineTruncationMiddle
    default: return kCTLineTruncationStart
    }
}

@_cdecl("ct_line_create_with_attributed_string")
func ct_line_create_with_attributed_string(
    _ attributedStringPtr: UnsafeMutableRawPointer?
) -> UnsafeMutableRawPointer? {
    guard let attributedStringPtr else { return nil }
    let attributedString: NSAttributedString = unbox(attributedStringPtr, as: NSAttributedString.self)
    return retainBox(CTLineCreateWithAttributedString(attributedString))
}

@_cdecl("ct_line_create_truncated_line")
func ct_line_create_truncated_line(
    _ linePtr: UnsafeMutableRawPointer?,
    _ width: Double,
    _ truncationTypeRaw: UInt32,
    _ truncationTokenPtr: UnsafeMutableRawPointer?
) -> UnsafeMutableRawPointer? {
    guard let linePtr else { return nil }
    let line: CTLine = unbox(linePtr, as: CTLine.self)
    let truncationToken: CTLine? = truncationTokenPtr.map { unbox($0, as: CTLine.self) }
    guard let truncated = CTLineCreateTruncatedLine(line, width, truncationType(truncationTypeRaw), truncationToken) else {
        return nil
    }
    return retainBox(truncated)
}

@_cdecl("ct_line_create_justified_line")
func ct_line_create_justified_line(
    _ linePtr: UnsafeMutableRawPointer?,
    _ justificationFactor: Double,
    _ justificationWidth: Double
) -> UnsafeMutableRawPointer? {
    guard let linePtr else { return nil }
    let line: CTLine = unbox(linePtr, as: CTLine.self)
    guard let justified = CTLineCreateJustifiedLine(line, CGFloat(justificationFactor), justificationWidth) else {
        return nil
    }
    return retainBox(justified)
}

@_cdecl("ct_line_get_glyph_count")
func ct_line_get_glyph_count(_ linePtr: UnsafeMutableRawPointer?) -> Int {
    guard let linePtr else { return 0 }
    let line: CTLine = unbox(linePtr, as: CTLine.self)
    return CTLineGetGlyphCount(line)
}

@_cdecl("ct_line_get_string_range")
func ct_line_get_string_range(_ linePtr: UnsafeMutableRawPointer?) -> CFRange {
    guard let linePtr else { return CFRange(location: 0, length: 0) }
    let line: CTLine = unbox(linePtr, as: CTLine.self)
    return CTLineGetStringRange(line)
}

@_cdecl("ct_line_get_pen_offset_for_flush")
func ct_line_get_pen_offset_for_flush(
    _ linePtr: UnsafeMutableRawPointer?,
    _ flushFactor: Double,
    _ flushWidth: Double
) -> Double {
    guard let linePtr else { return 0 }
    let line: CTLine = unbox(linePtr, as: CTLine.self)
    return CTLineGetPenOffsetForFlush(line, CGFloat(flushFactor), flushWidth)
}

@_cdecl("ct_line_get_typographic_bounds")
func ct_line_get_typographic_bounds(
    _ linePtr: UnsafeMutableRawPointer?,
    _ ascent: UnsafeMutablePointer<CGFloat>?,
    _ descent: UnsafeMutablePointer<CGFloat>?,
    _ leading: UnsafeMutablePointer<CGFloat>?
) -> Double {
    guard let linePtr else { return 0 }
    let line: CTLine = unbox(linePtr, as: CTLine.self)
    return CTLineGetTypographicBounds(line, ascent, descent, leading)
}

@_cdecl("ct_line_get_bounds_with_options")
func ct_line_get_bounds_with_options(_ linePtr: UnsafeMutableRawPointer?, _ options: UInt64) -> CGRect {
    guard let linePtr else { return .null }
    let line: CTLine = unbox(linePtr, as: CTLine.self)
    return CTLineGetBoundsWithOptions(line, CTLineBoundsOptions(rawValue: UInt(options)))
}

@_cdecl("ct_line_get_trailing_whitespace_width")
func ct_line_get_trailing_whitespace_width(_ linePtr: UnsafeMutableRawPointer?) -> Double {
    guard let linePtr else { return 0 }
    let line: CTLine = unbox(linePtr, as: CTLine.self)
    return CTLineGetTrailingWhitespaceWidth(line)
}

@_cdecl("ct_line_get_image_bounds")
func ct_line_get_image_bounds(_ linePtr: UnsafeMutableRawPointer?) -> CGRect {
    guard let linePtr else { return .null }
    let line: CTLine = unbox(linePtr, as: CTLine.self)
    return CTLineGetImageBounds(line, nil)
}

@_cdecl("ct_line_get_string_index_for_position")
func ct_line_get_string_index_for_position(
    _ linePtr: UnsafeMutableRawPointer?,
    _ position: CGPoint
) -> Int {
    guard let linePtr else { return kCFNotFound }
    let line: CTLine = unbox(linePtr, as: CTLine.self)
    return CTLineGetStringIndexForPosition(line, position)
}

@_cdecl("ct_line_get_offset_for_string_index")
func ct_line_get_offset_for_string_index(
    _ linePtr: UnsafeMutableRawPointer?,
    _ charIndex: Int,
    _ secondaryOffset: UnsafeMutablePointer<CGFloat>?
) -> CGFloat {
    guard let linePtr else { return 0 }
    let line: CTLine = unbox(linePtr, as: CTLine.self)
    return CTLineGetOffsetForStringIndex(line, charIndex, secondaryOffset)
}

@_cdecl("ct_line_get_run_count")
func ct_line_get_run_count(_ linePtr: UnsafeMutableRawPointer?) -> Int {
    guard let linePtr else { return 0 }
    let line: CTLine = unbox(linePtr, as: CTLine.self)
    let runs: [CTRun] = typedArray(CTLineGetGlyphRuns(line))
    return runs.count
}

@_cdecl("ct_line_copy_runs")
func ct_line_copy_runs(
    _ linePtr: UnsafeMutableRawPointer?,
    _ buffer: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ capacity: Int
) -> Int {
    guard let linePtr else { return 0 }
    let line: CTLine = unbox(linePtr, as: CTLine.self)
    let runs: [CTRun] = typedArray(CTLineGetGlyphRuns(line))
    return fillBoxedArray(runs, buffer: buffer, capacity: capacity)
}

File: src/line.rs
use crate::attributed_string::AttributedString;
use crate::bridge;
use crate::common::{expect_handle, impl_handle};
use crate::error::CoreTextResult;
use crate::run::CTRun;
use crate::types::{CGPoint, CGRect, TextRange, TypographicBounds};

/// Option flags for `CTLine::bounds_with_options`.
pub mod bounds_options {
    pub const EXCLUDE_TYPOGRAPHIC_LEADING: u64 = 1 << 0;
    pub const EXCLUDE_TYPOGRAPHIC_SHIFTS: u64 = 1 << 1;
    pub const USE_HANGING_PUNCTUATION: u64 = 1 << 2;
    pub const USE_GLYPH_PATH_BOUNDS: u64 = 1 << 3;
    pub const USE_OPTICAL_BOUNDS: u64 = 1 << 4;
    pub const INCLUDE_LANGUAGE_EXTENTS: u64 = 1 << 5;
}

/// Truncation strategy for `CTLineCreateTruncatedLine`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
#[repr(u32)]
pub enum LineTruncationType {
    #[default]
    Start = 0,
    End = 1,
    Middle = 2,
}

/// An immutable `CTLine` wrapper.
pub struct CTLine {
    raw: bridge::Handle,
}

impl_handle!(CTLine);

impl CTLine {
    pub fn create_with_attributed_string(
        attributed_string: &AttributedString,
    ) -> CoreTextResult<Self> {
        let raw =
            unsafe { bridge::ct_line_create_with_attributed_string(attributed_string.as_raw()) };
        Ok(Self::from_raw(expect_handle(
            raw,
            "ct_line_create_with_attributed_string returned NULL",
        )?))
    }

    #[must_use]
    pub fn truncated(
        &self,
        width: f64,
        truncation_type: LineTruncationType,
        truncation_token: Option<&Self>,
    ) -> Option<Self> {
        let raw = unsafe {
            bridge::ct_line_create_truncated_line(
                self.raw,
                width,
                truncation_type as u32,
                truncation_token.map_or(std::ptr::null_mut(), Self::as_raw),
            )
        };
        if raw.is_null() {
            None
        } else {
            Some(Self::from_raw(raw))
        }
    }

    #[must_use]
    pub fn justified(&self, justification_factor: f64, justification_width: f64) -> Option<Self> {
        let raw = unsafe {
            bridge::ct_line_create_justified_line(
                self.raw,
                justification_factor,
                justification_width,
            )
        };
        if raw.is_null() {
            None
        } else {
            Some(Self::from_raw(raw))
        }
    }

    #[must_use]
    pub fn glyph_count(&self) -> isize {
        unsafe { bridge::ct_line_get_glyph_count(self.raw) }
    }

    #[must_use]
    pub fn string_range(&self) -> TextRange {
        unsafe { bridge::ct_line_get_string_range(self.raw) }.into()
    }

    #[must_use]
    pub fn pen_offset_for_flush(&self, flush_factor: f64, flush_width: f64) -> f64 {
        unsafe { bridge::ct_line_get_pen_offset_for_flush(self.raw, flush_factor, flush_width) }
    }

    #[must_use]
    pub fn typographic_bounds(&self) -> TypographicBounds {
        let mut ascent = 0.0;
        let mut descent = 0.0;
        let mut leading = 0.0;
        let width = unsafe {
            bridge::ct_line_get_typographic_bounds(
                self.raw,
                &mut ascent,
                &mut descent,
                &mut leading,
            )
        };
        TypographicBounds {
            width,
            ascent,
            descent,
            leading,
        }
    }

    #[must_use]
    pub fn bounds_with_options(&self, options: u64) -> CGRect {
        unsafe { bridge::ct_line_get_bounds_with_options(self.raw, options) }
    }

    #[must_use]
    pub fn trailing_whitespace_width(&self) -> f64 {
        unsafe { bridge::ct_line_get_trailing_whitespace_width(self.raw) }
    }

    #[must_use]
    pub fn image_bounds(&self) -> CGRect {
        unsafe { bridge::ct_line_get_image_bounds(self.raw) }
    }

    #[must_use]
    pub fn string_index_for_position(&self, position: CGPoint) -> isize {
        unsafe { bridge::ct_line_get_string_index_for_position(self.raw, position) }
    }

    #[must_use]
    pub fn offset_for_string_index(&self, char_index: isize) -> (f64, f64) {
        let mut secondary = 0.0;
        let primary = unsafe {
            bridge::ct_line_get_offset_for_string_index(self.raw, char_index, &mut secondary)
        };
        (primary, secondary)
    }

    #[must_use]
    pub fn runs(&self) -> Vec<CTRun> {
        let count = unsafe { bridge::ct_line_get_run_count(self.raw) };
        if count <= 0 {
            return Vec::new();
        }
        let mut handles = vec![std::ptr::null_mut(); usize::try_from(count).unwrap_or(0)];
        let written = unsafe { bridge::ct_line_copy_runs(self.raw, handles.as_mut_ptr(), count) };
        handles.truncate(usize::try_from(written).unwrap_or(0));
        handles.into_iter().map(CTRun::from_raw).collect()
    }
}

#[must_use]
pub fn line_type_id() -> u64 {
    unsafe { bridge::ct_line_get_type_id() }
}
